provisioning.go

```diff
@@ -0,0 +1,370 @@
+	ProvisionModeClear ProvisionMode = iota
+
+	// ProvisionModeWithoutLockout specifies that the TPM should be refreshed without performing
+	// operations that require knowledge of the lockout hierarchy authorization value.
+	ProvisionModeWithoutLockout
+
+	// ProvisionModeFull specifies that the TPM should be fully provisioned without clearing it.
+	ProvisionModeFull
+)
+
+var (
+	// srkTemplate is the default RSA2048 SRK template, see section 7.5.1 of "TCG TPM v2.0 Provisioning Guidance", version 1.0,
+	// revision 1.0, 15 March 2017.
+	srkTemplate = tpm2.Public{
+		Type:    tpm2.ObjectTypeRSA,
+		NameAlg: tpm2.HashAlgorithmSHA256,
+		Attrs: tpm2.AttrFixedTPM | tpm2.AttrFixedParent | tpm2.AttrSensitiveDataOrigin | tpm2.AttrUserWithAuth | tpm2.AttrNoDA |
+			tpm2.AttrRestricted | tpm2.AttrDecrypt,
+		Params: tpm2.PublicParamsU{
+			Data: &tpm2.RSAParams{
+				Symmetric: tpm2.SymDefObject{
+					Algorithm: tpm2.SymObjectAlgorithmAES,
+					KeyBits:   tpm2.SymKeyBitsU{Data: uint16(128)},
+					Mode:      tpm2.SymModeU{Data: tpm2.SymModeCFB}},
+				Scheme:   tpm2.RSAScheme{Scheme: tpm2.RSASchemeNull},
+				KeyBits:  2048,
+				Exponent: 0}},
+		Unique: tpm2.PublicIDU{Data: make(tpm2.PublicKeyRSA, 256)}}
+)
+
+// ProvisionTPM prepares the TPM associated with the tpm parameter for full disk encryption. The mode parameter specifies the
+// behaviour of this function.
+//
+// If mode is not ProvisionModeWithoutLockout, this function performs operations that require knowledge of the lockout hierarchy
+// authorization value, which must be provided by calling TPMConnection.LockoutHandleContext().SetAuthValue() prior to this call.
+// If the wrong lockout hierarchy authorization value is provided, then a AuthFailError error will be returned. If this happens,
+// the TPM will have entered dictionary attack lockout mode for the lockout hierarchy. Further calls will result in a ErrTPMLockout
+// error being returned. The only way to recover from this is to either wait for the pre-programmed recovery time to expire, or to
+// clear the TPM via the physical presence interface.
+//
+// If mode is ProvisionModeClear, this function will attempt to clear the TPM before provisioning it. If owner clear has been disabled
+// (which will be the case if the TPM has previously been provisioned with this function), then ErrTPMClearRequiresPPI will be returned.
+// In this case, the TPM must be cleared via the physical presence interface by calling RequestTPMClearUsingPPI and performing a
+// system restart.
+//
+// This function will create and persist an endorsement key which requires knowledge of the authorization values for the storage
+// and endorsement hierarchies. If called with mode set to ProvisionModeClear, or if called just after clearing the TPM via the
+// physical presence interface, the authorization values for these hierarchies will be empty at the point that they are required.
```

Comment:
is this a TPM termology thing to use empty vs unset,  is it actually conceptual?  (I'm asking because unset would be a bit natural to me)

Reply:
The TCG specifications refer to an unset authorization value as empty.